### DISCUSSION TITLE
Intellisense support for editing Mill script

### DISCUSSION THREAD

**Discussion by hmf** [scalameta/metals#5536] [Q&A] [answered]
I am trying to reach developer "nirvana" by using Metals to edit and develop my Mills scripts. Currently I have to replicate the project and use IntelliJ for this. In my experiments I found that because the *.sc are treated as Ammonite scripts by Metals, I do not have access to the Mill API. 

I tried then to added the following imports:

```scala
import $ivy.`com.lihaoyi:mill-main_2.13:0.11.1`
import $ivy.`com.lihaoyi:mill-scalalib_2.13:0.11.1`
import $ivy.`com.lihaoyi:mill-contrib-bloop_2.13:0.11.1`
```

And placed in the first line of the script:
```scala
// scala 2.13.3
```

Now intellisense works well, however the following code will fail to compile:

```scala
import mill._
import mill.scalalib._
import mill.define.ModuleRef
import coursier.maven.MavenRepository
import mill.contrib.bloop.Bloop

val ScalaVersion = "3.3.0"


trait CommonSettings extends SbtModule with Bloop.Module {
  def scalaVersion = T{ ScalaVersion }

  def aTask = T {
    T.log.debug("This works")
  }
}

object docs extends CommonSettings {
}
```

For the `CommonSettings` in the `docs` object I get the IDE error message:
 
```shell
 could not find implicit value for parameter outerCtx0: mill.define.Ctxscalac
```

Seems like Mill's macros don't work here. My other strategy would be to substitute the ammonite binary with the `millw` one. Is this a possible solution? If so, were do I set the Ammonite path? Could not find the answer when using the documentation site.

TIA

**Comment by tgodzik** [comment, marked as answer]
Have you tried using mill via BSP for this? Ideally this would be supported as a separate build target there, but I don't exactly remember if that is currently the case. Alternatively, we could go the sbt Bloop plugin way, where we would generate additional Bloop config for mill build definition when running the import via `mill.contrib.bloop.Bloop/install` 

For the second option I don't have any time to work on that currently unfortunately, so this would need some love from anyone with spare time on their hands.

**Reply by hmf** [reply]
I am using mill via BSP but for a standard project. I don't understand what you mean by  "a separate build target ". Do you mean add the script as if it were a source file? 

Thanks for the feedback.

**Reply by tgodzik** [reply]
I would expect mill BSP to report the build scripts as a separate compilation target, but it might not be the case then :thinking:  With sbt we have a separate target for each sbt meta build, which has all the info that we need like classpath and dependencies and actual sources to take into account, Without that info we can't support build scripts reliably and there is no way currently to differentiate script and mill source if they are not named `build.sc`

**Reply by hmf** [reply]
> I would expect mill BSP to report the build scripts as a separate compilation target,

Understood. Don't think is possible automatically. I will keep this in mind in case I get some other idea.

>  ... there is no way currently to differentiate script and mill source if they are not named build.sc

True. I also tried this as a worksheet, but had compilation errors. 

Thank you.

**Comment by hmf** [comment]
I have my answer. I will assume this is not possible at the moment so am closing this thread.
